<h1><b>Tópicos Especiais: Aprendizado de Máquina</b><h1>
<H3>Prof. Alex Sandro</H3>
<h3>Curso Superior de Tecnologia em Sistemas para Internet</h3>



In [ ]:
# consultando a versão do Python instalado na VM do Colab
!python -V

Python 3.11.11


**Imports**

In [ ]:
import pandas as pd
import numpy as np

<H1><b>📘 Introdução</b></h1>

Este notebook contém uma relação de tarefas usualmente executadas durante a etapa de **pré-processamento** do pipeline de ciência de dados.

# 1.Carga dos Dados

##1.1 A partir de um link no `GitHub`

**TMDB 5000 Movie Dataset**:
https://www.kaggle.com/datasets/tmdb/tmdb-movie-metadata

In [ ]:
import pandas as pd

# Caminho completo para o arquivo
file_path = 'https://raw.githubusercontent.com/alex-ifpb/datasets/refs/heads/main/ciencia_de_dados/tmdb_5000_movies.csv'

# Carregando o dataset
df_filmes = pd.read_csv(file_path)

# Exibindo as primeiras linhas
df_filmes.head(3)


,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...","[{""iso_3166_1"": ""GB"", ""name"": ""United Kingdom""...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466


##1.2 Upload de arquivo externo

Nesta opção, acesse o menu à esquerda do notebook e procure o ícone *Arquivos*. Faça o upload do arquivo desejado. Lembrando que, ao encerrar a sessão com o Google Colab, os arquivos "subidos" serão descartados.

In [ ]:
import pandas as pd
# import seaborn as sns
# import matplotlib.pyplot as plt
import numpy as np
df_filmes  = pd.read_csv('tmdb_5000_movies.csv')

df_ratings = pd.read_csv('ratings.csv', sep=',')
df_ratings.columns = ["usuarioId", "filmeId", "nota","momento"]

df_movies = pd.read_csv('movies.csv', sep=',',)
df_movies.columns = ['filmeId','titulo','genero']

##1.3 Carregando a partir do `Google Drive`

Nesta opção, o drive do google é montado para uso no ambiente Colaboratoy. Se o notebook for compartilhado, o usuário final não poderá "montar" o drive da conta de terceiros.

In [ ]:
from google.colab import drive
import pandas as pd

drive.mount('/content/drive', force_remount=True)

# Caminho completo para o arquivo
file_path = '/content/drive/My Drive/datasets/tmdb_5000_movies.csv'

# Carregando o dataset
df_filmes = pd.read_csv(file_path)

# Exibindo as primeiras linhas
df_filmes.head(3)

#2.Explorando o dataset

##2.1 Metadados e estatísticas

In [ ]:
df_filmes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

In [ ]:
df_filmes.describe()

statistic,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
str,f64,str,str,f64,str,str,str,str,f64,str,str,str,f64,f64,str,str,str,str,f64,f64
"""count""",4803.0,"""4803""","""1712""",4803.0,"""4803""","""4803""","""4803""","""4800""",4803.0,"""4803""","""4803""","""4802""",4803.0,4801.0,"""4803""","""4803""","""3959""","""4803""",4803.0,4803.0
"""null_count""",0.0,"""0""","""3091""",0.0,"""0""","""0""","""0""","""3""",0.0,"""0""","""0""","""1""",0.0,2.0,"""0""","""0""","""844""","""0""",0.0,0.0
"""mean""",2.9045e7,null,null,57165.484281,null,null,null,null,21.492301,null,null,null,8.2261e7,106.875859,null,null,null,null,6.092172,690.217989
"""std""",4.0722e7,null,null,88694.614033,null,null,null,null,31.81665,null,null,null,1.6286e8,22.611935,null,null,null,null,1.194612,1234.585891
"""min""",0.0,"""[]""","""http://2016themovie.com/""",5.0,"""[]""","""af""","""#Horror""",""" """,0.0,"""[]""","""[]""","""1916-09-04""",0.0,0.0,"""[]""","""Post Production""","""""All the world likes an outlaw…","""#Horror""",0.0,0.0
"""25%""",800000.0,null,null,9016.0,null,null,null,null,4.66891,null,null,null,0.0,94.0,null,null,null,null,5.6,54.0
"""50%""",1.5e7,null,null,14629.0,null,null,null,null,12.921594,null,null,null,1.9170001e7,103.0,null,null,null,null,6.2,235.0
"""75%""",4e7,null,null,58626.0,null,null,null,null,28.350529,null,null,null,9.2921203e7,118.0,null,null,null,null,6.8,737.0
"""max""",3.8e8,"""[{""id"": 99, ""name"": ""Documenta…","""https://www.warnerbros.com/wat…",459488.0,"""[{""id"": 999, ""name"": ""sports c…","""zh""","""해운대""","""“The Perfect Wave” is the true…",875.581305,"""[{""name"": ""rusty bear entertai…","""[{""iso_3166_1"": ""ZA"", ""name"": …","""2017-02-03""",2.7880e9,338.0,"""[{""iso_639_1"": ""zh"", ""name"": ""…","""Rumored""","""最后的舞者""","""Æon Flux""",10.0,13752.0


In [ ]:
# Número de linhas e atributos
df_filmes.shape

(4803, 20)

##2.2 Testando algumas operações úteis

###2.2.1 unique()

In [ ]:
# variável categórica do atributo "original_language" sem repetição
df_filmes['original_language'].unique()
#df_filmes.original_language.unique()

original_language
str
"""xx"""
"""ta"""
"""ru"""
"""ps"""
"""vi"""
…
"""es"""
"""fa"""
"""is"""


###2.2.2 describe()

In [ ]:
df_filmes['vote_count'].describe()

statistic,value
str,f64
"""count""",4803.0
"""null_count""",0.0
"""mean""",690.217989
"""std""",1234.585891
"""min""",0.0
"""25%""",54.0
"""50%""",235.0
"""75%""",737.0
"""max""",13752.0


* **25% dos dados (Primeiro Quartil -> Q1) = 54.0**: Significa que 25% das produções têm até 54 votos. É a cauda inferior, onde estão os itens menos votados
*   **50% dos dados (Mediana -> Q2) = 235.0**: É o centro exato. Metade das produções (50%) recebeu menos de 235 votos, e a outra metade recebeu mais do que isso
*   **75% dos dados (Terceiro Quartil -> Q3) = 737.0**: Significa que 75% das produções acumulam até 737 votos. Apenas os 25% mais populares ultrapassam esse patamar.
* O item mais popular do conjunto recebeu uma concentração de 13.752 votos.

###2.2.3 value_counts()

In [ ]:
df_filmes['original_language'].value_counts().head()

original_language,count
str,u32
"""fa""",4
"""ru""",11
"""ja""",16
"""zh""",27
"""cs""",2


In [ ]:
# variável quantitativa contínua
df_filmes['budget']

budget
i64
237000000
300000000
245000000
250000000
260000000
…
220000
9000
0


In [ ]:
# variável quantitativa: vote_count, popularity
df_filmes['vote_count']

vote_count
i64
11800
4500
4466
9106
2124
…
238
5
6


#3.Imputação de Valores

##3.1 Preenchimento de valores ausentes com a **média**

In [ ]:
# Criar um DataFrame de exemplo com valores ausentes (NaN)
import pandas as pd
import numpy as np
data = {'janeiro': [1, 2, np.nan, 4, 5],
        'fevereiro': [np.nan, 6, 7, 8, np.nan],
        'março': [10, 11, 12, np.nan, 14]}
df = pd.DataFrame(data)

print("\nNúmero de Veiculos Vendidos")
print(df.isnull())

print("\nDataFrame original:")
print("Número de Veiculos Vendidos")
print(df)
print("\n")

# Calcular a média de cada coluna
media_coluna_jan = df['janeiro'].mean()
media_coluna_fev = df['fevereiro'].mean()
media_coluna_mar = df['março'].mean()

print(f"Média da coluna A: {media_coluna_jan:.2f}")
print(f"Média da coluna B: {media_coluna_fev:.2f}")
print(f"Média da coluna C: {media_coluna_mar:.2f}")
print("\n")

# Preencher os valores ausentes na coluna 'A' com sua média
# df['janeiro'].fillna(media_coluna_jan, inplace=True)
df.fillna({'janeiro':media_coluna_jan}, inplace=True)

# Preencher os valores ausentes na coluna 'B' com sua média
# df['fevereiro'].fillna(media_coluna_fev, inplace=True)
df['fevereiro'] = df['fevereiro'].fillna(media_coluna_fev)

# Preencher os valores ausentes na coluna 'C' com sua média
# df['março'].fillna(media_coluna_mar, inplace=True)
df.fillna({'março':media_coluna_mar}, inplace=True)

print("DataFrame após preencher os valores ausentes com a média de cada coluna:")
print("\nNúmero de Veiculos Vendidos")
print(df)
print()



Número de Veiculos Vendidos
   janeiro  fevereiro  março
0    False       True  False
1    False      False  False
2     True      False  False
3    False      False   True
4    False       True  False

DataFrame original:
Número de Veiculos Vendidos
   janeiro  fevereiro  março
0      1.0        NaN   10.0
1      2.0        6.0   11.0
2      NaN        7.0   12.0
3      4.0        8.0    NaN
4      5.0        NaN   14.0


Média da coluna A: 3.00
Média da coluna B: 7.00
Média da coluna C: 11.75


DataFrame após preencher os valores ausentes com a média de cada coluna:

Número de Veiculos Vendidos
   janeiro  fevereiro  março
0      1.0        7.0  10.00
1      2.0        6.0  11.00
2      3.0        7.0  12.00
3      4.0        8.0  11.75
4      5.0        7.0  14.00



##2.2 Preenchimento pela **mediana**

In [ ]:
import pandas as pd
import numpy as np

# Criando um DataFrame com valores ausentes
dados = {
    'idade': [25, 30, np.nan, 22, 28, np.nan, 60]
}

df = pd.DataFrame(dados)

print("Antes do tratamento:")
print(df)

# Substituindo os valores ausentes pela mediana da coluna
mediana_idade = df['idade'].median()

print('\nMediana = ', mediana_idade)
# df['idade'].fillna(mediana_idade, inplace=True)
df.fillna({'idade':mediana_idade}, inplace=True)

print("\nDepois do tratamento:")
print(df)

Antes do tratamento:
   idade
0   25.0
1   30.0
2    NaN
3   22.0
4   28.0
5    NaN
6   60.0

Mediana =  28.0

Depois do tratamento:
   idade
0   25.0
1   30.0
2   28.0
3   22.0
4   28.0
5   28.0
6   60.0


##2.3 Preenchimento com a moda

In [ ]:
import pandas as pd
import numpy as np

# Criar um DataFrame de exemplo com valores ausentes em um atributo categórico
data = {'ID Cliente': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
        'Cor Favorita': ['Azul', 'Verde', 'Azul', np.nan, 'Vermelho',
                           'Azul', 'Amarelo', 'Azul', np.nan, 'Azul']}
df = pd.DataFrame(data)

print("DataFrame original:")
print(df)
print("\n")

# Calcular a moda da coluna 'Cor Favorita'
moda_cor = df['Cor Favorita'].mode()[0]  # .mode() retorna uma Series, pegamos o primeiro elemento

print(f"Moda da coluna 'Cor Favorita': {moda_cor}")
print("\n")

# Preencher os valores ausentes na coluna 'Cor Favorita' com a moda
df['Cor Favorita'].fillna(moda_cor, inplace=True)

print("DataFrame após preencher os valores ausentes com a moda:")
print(df)



DataFrame original:
   ID Cliente Cor Favorita
0           1         Azul
1           2        Verde
2           3         Azul
3           4          NaN
4           5     Vermelho
5           6         Azul
6           7      Amarelo
7           8         Azul
8           9          NaN
9          10         Azul


Moda da coluna 'Cor Favorita': Azul


DataFrame após preencher os valores ausentes com a moda:
   ID Cliente Cor Favorita
0           1         Azul
1           2        Verde
2           3         Azul
3           4         Azul
4           5     Vermelho
5           6         Azul
6           7      Amarelo
7           8         Azul
8           9         Azul
9          10         Azul


<ipython-input-13-7cbcf7e9c6ac>:21: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Cor Favorita'].fillna(moda_cor, inplace=True)


##2.4 Interpolação (séries temporais): preenchimento com o **anterior/posterior**


In [ ]:
import pandas as pd
import numpy as np

# Exemplo de dados temporais com valores ausentes
data = {
    'data': pd.date_range('2024-01-01', periods=7),
    'temperatura': [22.1, np.nan, np.nan, 23.0, np.nan, 24.5, np.nan]
}
df = pd.DataFrame(data)
print()
print(df.info())
print()

print("Original:")
print(df)

# Preenchimento com valor anterior (forward fill)
print("\nForward Fill (valor anterior):")
df_ffill = df.copy()
# df_ffill['temperatura'] = df_ffill['temperatura'].fillna(method='ffill')
df_ffill['temperatura'] = df_ffill['temperatura'].ffill()
print(df_ffill)

# Preenchimento com valor posterior (backward fill)
print("\nBackward Fill (valor posterior):")
df_bfill = df.copy()
# df_bfill['temperatura'] = df_bfill['temperatura'].fillna(method='bfill')
df_bfill['temperatura'] = df_bfill['temperatura'].bfill()
print(df_bfill)






<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   data         7 non-null      datetime64[ns]
 1   temperatura  3 non-null      float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 244.0 bytes
None

Original:
        data  temperatura
0 2024-01-01         22.1
1 2024-01-02          NaN
2 2024-01-03          NaN
3 2024-01-04         23.0
4 2024-01-05          NaN
5 2024-01-06         24.5
6 2024-01-07          NaN

Forward Fill (valor anterior):
        data  temperatura
0 2024-01-01         22.1
1 2024-01-02         22.1
2 2024-01-03         22.1
3 2024-01-04         23.0
4 2024-01-05         23.0
5 2024-01-06         24.5
6 2024-01-07         24.5

Backward Fill (valor posterior):
        data  temperatura
0 2024-01-01         22.1
1 2024-01-02         23.0
2 2024-01-03         23.0
3 2024-01-04         23.0
4 2024-01-0

#4.Descobrindo Valores Nulos

##4.1 a função `info()` do dataframe

In [ ]:
df_filmes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

##4.2 A função `isnull()`

In [ ]:
df_filmes.isnull().tail()

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
4798,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4799,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4800,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
4801,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4802,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False


In [ ]:
print('Homepage (#nulos):', df_filmes.homepage.isnull().sum())
# mostrando os nulos
nulos = df_filmes.homepage.isnull() # matriz de nulos
df_filmes[nulos].head(5)

Homepage (#nulos): 3091


,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
15,225000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 10751...",NaN,2454,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...",en,The Chronicles of Narnia: Prince Caspian,One year after their incredible adventures in ...,53.978602,"[{""name"": ""Walt Disney"", ""id"": 5888}, {""name"":...","[{""iso_3166_1"": ""CZ"", ""name"": ""Czech Republic""...",2008-05-15,419651413,150.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,Hope has a new face.,The Chronicles of Narnia: Prince Caspian,6.3,1630
24,207000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 18, ""...",NaN,254,"[{""id"": 774, ""name"": ""film business""}, {""id"": ...",en,King Kong,"In 1933 New York, an overly ambitious movie pr...",61.226010,"[{""name"": ""WingNut Films"", ""id"": 11}, {""name"":...","[{""iso_3166_1"": ""NZ"", ""name"": ""New Zealand""}, ...",2005-12-14,550000000,187.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The eighth wonder of the world.,King Kong,6.6,2337
27,209000000,"[{""id"": 53, ""name"": ""Thriller""}, {""id"": 28, ""n...",NaN,44833,"[{""id"": 1721, ""name"": ""fight""}, {""id"": 4410, ""...",en,Battleship,"When mankind beams a radio signal into space, ...",64.928382,"[{""name"": ""Universal Pictures"", ""id"": 33}, {""n...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-04-11,303025485,131.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,The Battle for Earth Begins at Sea,Battleship,5.5,2114
33,210000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 28, ""...",NaN,36668,"[{""id"": 1852, ""name"": ""mutant""}, {""id"": 8828, ...",en,X-Men: The Last Stand,"When a cure is found to treat mutations, lines...",3.857526,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""CA"", ""name"": ""Canada""}, {""iso...",2006-05-24,459359555,104.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,Take a Stand,X-Men: The Last Stand,6.3,3525
34,200000000,"[{""id"": 16, ""name"": ""Animation""}, {""id"": 10751...",NaN,62211,"[{""id"": 1299, ""name"": ""monster""}, {""id"": 5984,...",en,Monsters University,A look at the relationship between Mike and Su...,89.186492,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2013-06-20,743559607,104.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,School never looked this scary.,Monsters University,7.0,3528


In [ ]:
df_filmes.isnull().sum()

,0
budget,0
genres,0
homepage,3091
id,0
keywords,0
original_language,0
original_title,0
overview,3
popularity,0
production_companies,0


##4.3 `fillna()`

In [ ]:
df = pd.DataFrame({
    'nome': ['Ana', 'Bruno', 'Carlos', 'Diana'],
    'profissao': ['Engenheira', None, 'Médico', None]
})

print("Antes do tratamento:")
print(df)

# Substituindo valores ausentes por uma categoria única
df['profissao'] = df['profissao'].fillna('Desconhecido')

print("\nDepois do tratamento:")
print(df)

Antes do tratamento:
     nome   profissao
0     Ana  Engenheira
1   Bruno        None
2  Carlos      Médico
3   Diana        None

Depois do tratamento:
     nome     profissao
0     Ana    Engenheira
1   Bruno  Desconhecido
2  Carlos        Médico
3   Diana  Desconhecido


# 5.Transformação de Dados (Parte 1)

A **transformação de dados** é uma etapa do pré-processamento em que as **variáveis originais** são **modificadas**, **reescalonadas** ou **convertidas** para uma nova representação mais adequada à análise ou ao modelo de aprendizado de máquina.

##5.1 Discretização

A **discretização** consiste em transformar **variáveis numéricas contínuas** (como idade, salário, temperatura etc.) em **variáveis categóricas** (como faixas etárias, níveis salariais, categorias de temperatura etc.).

Ela simplifica grandes conjuntos de valores contínuos em **intervalos gerenciáveis**, reduzindo assim a complexidade.

Um dos métodos comuns de discretização é o **binning** (agrupamento em "caixas"), técnica que divide os dados **em intervalos fixos (equal-width binning)** e baseado na frequência de elementos (**equal-frequency binning**).

In [ ]:
import pandas as pd

# Dados fictícios
dados = pd.DataFrame({
    'idade': [15, 22, 37, 45, 52, 60, 67, 80, 90]
})

**Vantagens:**

* **Redução de Ruído:** Ajuda a mitigar o efeito de *outliers* (valores discrepantes), colocando-os em intervalos extremos.
* **Melhora a interpretação dos dados**: facilita a compreensão dos resultados por humanos (ex: "faixa etária" é mais interpretável do que "idade exata").
* **Otimiza o processamento**: algoritmos podem computar mais rápido com menos valores distintos

**Desvantagens:**

* **Perda de Informação e Precisão**: Ao agrupar valores distintos em um mesmo intervalo (bin), perde-se a variação detalhada dentro de cada grupo. Valores que eram sutilmente diferentes, mas importantes para a predição, tornam-se idênticos
* **Perda de Poder Estatístico**: A conversão de dados contínuos para categóricos reduz a capacidade do modelo de identificar tendências lineares ou relações precisas.
* **Incompatibilidade com Algoritmos**: Alguns modelos de aprendizado de máquina, especialmente algoritmos baseados em regressão, funcionam melhor com dados contínuos e podem ter desempenho inferior com dados discretizados.
* **Descontinuidade**: A discretização transforma relações contínuas em "saltos" (jumps) entre os intervalos, o que pode representar mal a realidade de um fenômeno que muda de forma contínua

###5.1.1 Equal-width binning (intervalo fixo)

📘 **Separação Automática Conforme Quantidade de Bins**


No exemplo a seguir, **informamos a quantidade de bins**, para que a função `pd.cut()` **faça a divisão dos dados automaticamente**.

O `pd.cut()` com "equal-width binning" divide uma variável contínua em um **número predeterminado de intervalos de largura igual**, onde a função calcula o intervalo de valores (máximo - mínimo), o divide pelo número de bins especificado para obter a largura de cada bin, e então atribui cada valor a um desses intervalos.

In [ ]:
import pandas as pd
import numpy as np

valores = [5, 6, 5, 7, 12, 17, 18, 24, 33, 45, 59, 63, 70]
df = pd.DataFrame({'idade': valores})
print('Shape:', df.shape)

# 1. Equal-width binning (3 faixas de mesma largura)
df['idade_bin_width'] = pd.cut(df['idade'], bins=3,
                               labels=["Jovem","Meia-idade","Idoso"],)
print("\n=== Equal-width Binning ===")
print(df[['idade', 'idade_bin_width']])

print(f'\nFaixas de valores dos bins:')
print(df['idade_bin_width'].value_counts())

print('''
intervalo de valores de cada bin
bin 1 =  5 a 24 Jovem
bin 2 = 25 a 48 Meia-idade
bin 3 = 49 a 70 Idoso ''')

bin_series = pd.cut(df['idade'], bins=3,
                               labels=["Jovem","Meia-idade","Idoso"],)

# 1.1 Ver os intervalos de cada valor
#print(bin_series)
#print()
# 1.2 Ver apenas os intervalos únicos calculados
print(bin_series.unique())

Shape: (13, 1)

=== Equal-width Binning ===
    idade idade_bin_width
0       5           Jovem
1       6           Jovem
2       5           Jovem
3       7           Jovem
4      12           Jovem
5      17           Jovem
6      18           Jovem
7      24           Jovem
8      33      Meia-idade
9      45      Meia-idade
10     59           Idoso
11     63           Idoso
12     70           Idoso

Faixas de valores dos bins:
idade_bin_width
Jovem         8
Idoso         3
Meia-idade    2
Name: count, dtype: int64

intervalo de valores de cada bin
bin 1 =  5 a 24 Jovem
bin 2 = 25 a 48 Meia-idade
bin 3 = 49 a 70 Idoso 
['Jovem', 'Meia-idade', 'Idoso']
Categories (3, object): ['Jovem' < 'Meia-idade' < 'Idoso']


A explicação para o resultado é a seguinte:

1. Calcula o valor mínimo e máximo dos dados:

```
min = 5
max = 70
```

2. Cria 3 intervalos de mesmo tamanho:

Queremos 3 bins iguais, então:

Largura do bin = ((max - min)/nº bins)
```
Range total = 70 - 5 = 65
Tamanho de cada bin = 65 / 3 = 21.66
```

3. Intervalos criados serão:

```
Bin 1: (4.935, 26.66]   # 5, 5+21.66
Bin 2: (26.66, 48.33]
Bin 3: (48.33, 70]  # ultimo valor
```

Obs: `pd.cut()` usa uma pequena margem abaixo do mínimo para garantir que o menor valor caia no primeiro intervalo, por isso o início é 4.935 (ligeiramente abaixo de 5).

<p>

**Desvantagem**

Uma desvantagem do binning de largura fixa está relacionada a situações em que os dados **não estão uniformemente distribuídos**.

Como os bins são baseados em **intervalos iguais** de valores e não na **frequência dos dados**, você pode acabar com:

* Bins com muitos dados
* Bins quase vazios ou até vazios

Com `pd.cut(valores, bins=3)`, quase todos os valores ficam no primeiro bin:

* Bin 1: 5, 6, 5, 7, 12, 17, 18, 24 → 8 valores
* Bin 2: 33, 45 → 2 valores
* Bin 3: 59, 63, 70 → 3 valores

###5.1.2 Equal-frequency binning (quantis)

O **Equal-Frequency Binning** é um tipo de discretização em que os dados são divididos em **intervalos com número (frequência) aproximadamente igual de elementos**.

Diferente do `pd.cut()` (que divide os dados por valor), o `pd.qcut()` divide por quantidade de dados — ou seja, cria faixas com o **mesmo número de observações**.

O **quantil** é um termo geral para qualquer valor que **divide um conjunto de dados ordenados em partes iguais**, enquanto o **quartil** é um tipo específico de quantil que divide os dados em exatamente quatro partes

In [ ]:
# 2. Equal-frequency binning (quantis)
df['idade_bin_quantis'] = pd.qcut(df['idade'], q=3, labels=["Baixa",
                                                            "Média",
                                                            "Alta"])
print("\n=== Equal-frequency Binning ===")
print("            (Quantis)")
print(df[['idade', 'idade_bin_quantis']])


=== Equal-frequency Binning ===
            (Quantis)
    idade idade_bin_quantis
0       5             Baixa
1       6             Baixa
2       5             Baixa
3       7             Baixa
4      12             Média
5      17             Média
6      18             Média
7      24             Média
8      33              Alta
9      45              Alta
10     59              Alta
11     63              Alta
12     70              Alta


De acordo com o exmeplo, o `pd.qcut()` divide os dados ordenados em 3 grupos com número igual (ou quase igual) de observações:

Como temos 13 elementos, os grupos serão aproximadamente:

* Bin 1: 4 ou 5 elementos
* Bin 2: 4 ou 5 elementos
* Bin 3: 4 elementos

Internamente, ele usa os quantis 0%, 33.3%, 66.6%, 100% para dividir os dados.

O `pd.qcut(valores, q=3)` pode gerar algo como:

```
Intervalo	     Valores atribuídos
(4.999, 7.0]	 5,   5,  6, 7
(7.0, 24.0]	     12, 17, 18, 24
(24.0, 70.0]	 33, 45, 59, 63, 70
```

**Vantagens**
* Bins equilibrados
* Evita bins vazios ou desbalanceados

**Desvantagens**
* Pode quebrar faixas lógicas (Por exemplo, pode separar idades próximas em bins diferentes)
* Limites dos bins podem parecer estranhos (Podem sair com limites como 17.0 ou 45.0, que não têm significado real)

###5.1.3 Bins especificados por limites de valores


Esta técnica de discretização divide os dados de uma coluna **em faixas predefinidas**.

Considerando que desejamos transformar a variável "idade" (que pode ir de 0 a 100) em faixas tais como:

* 0–18: Jovem
* 19–59: Adulto
* 60+: Idoso

Note que, aqui, estaremos **especificando explicitamente** os limites de cada faixa usando uma lista. Isto será feito usando a função `pd.cut()`

In [ ]:
import pandas as pd

# Dados fictícios
dados = pd.DataFrame({
    'idade': [15, 18, 19, 22, 37, 45, 52, 60, 67, 80, 90]
})

bins = [0, 18, 59, 120]
labels = ['Jovem', 'Adulto', 'Idoso']

dados['idade_categoria'] = pd.cut(dados['idade'], bins=bins, labels=labels)

print(dados)

    idade idade_categoria
0      15           Jovem
1      18           Jovem
2      19          Adulto
3      22          Adulto
4      37          Adulto
5      45          Adulto
6      52          Adulto
7      60           Idoso
8      67           Idoso
9      80           Idoso
10     90           Idoso


###5.1.4 Discretização manual (regras do negócio)

Também chamada de **discretização supervisionada por domínio**, é um método em que desejamos transformar variáveis numéricas (como idade, renda ou temperatura) em **categorias discretas** que façam sentido no contexto do negócio ou da aplicação.

A **discretização manual** consiste em criar intervalos definidos pelo analista ou especialista de domínio, baseando-se em regras de negócio, conhecimento empírico ou critérios específicos.

Diferente das **técnicas automáticas** — como **equal-width binning** (largura igual) ou **equal-frequency binning** (frequência igual) —, essa abordagem **não depende apenas de estatísticas dos dados, mas sim de critérios interpretáveis e úteis para a tomada de decisão**.

In [ ]:
def faixa_idade(idade):
    if idade <= 12:
        return 'Criança'
    elif idade <= 17:
        return 'Adolescente'
    elif idade <= 59:
        return 'Adulto'
    else:
        return 'Idoso'

df['idade_manual'] = df['idade'].apply(faixa_idade)
print("\n=== Discretização Manual ===")
print(df[['idade', 'idade_manual']])


=== Discretização Manual ===
    idade idade_manual
0       5      Criança
1       6      Criança
2       5      Criança
3       7      Criança
4      12      Criança
5      17  Adolescente
6      18       Adulto
7      24       Adulto
8      33       Adulto
9      45       Adulto
10     59       Adulto
11     63        Idoso
12     70        Idoso


In [ ]:
import pandas as pd
import numpy as np

# Para reproduzir os mesmos resultados
np.random.seed(42)

# Gera 100 idades normalmente distribuídas entre aproximadamente 18 e 60 anos
idades = np.random.normal(loc=40, scale=8, size=100).astype(int)

# Insere alguns outliers
idades[5] = 5
idades[20] = 120
idades[45] = 3
idades[70] = 150

# Cria o DataFrame
df = pd.DataFrame({'idade': idades})

print(df)

    idade
0      43
1      38
2      45
3      52
4      38
..    ...
95     28
96     42
97     42
98     40
99     38

[100 rows x 1 columns]
